src/java_bytecode/java_bytecode_convert_class.cpp

```diff
@@ -317,20 +393,114 @@ void java_bytecode_convert_classt::convert(const classt &c)
     throw 0;
   }
 
-  // now do fields
+  // Now do fields
+  const class_typet::componentst &fields =
+    to_class_type(class_symbol->type).components();
+  // Include fields from overlay classes as they will be required by the
+  // methods defined there
+  for(auto overlay_class : overlay_classes)
+  {
+    for(const auto &field : overlay_class.get().fields)
+    {
+      std::string field_id = qualified_classname + "." + id2string(field.name);
+      if(check_field_exists(field, field_id, fields))
+      {
+        // TODO: This could be a warning if the types match
+        error()
+          << "Duplicate field definition for " << field_id
+          << " in overlay class" << eom;
+        continue;
+      }
+      debug()
+        << "Adding symbol from overlay class:  field '" << field.name << "'"
+        << eom;
+      convert(*class_symbol, field);
+      POSTCONDITION(check_field_exists(field, field_id, fields));
+    }
+  }
   for(const auto &field : c.fields)
   {
+    std::string field_id = qualified_classname + "." + id2string(field.name);
+    if(check_field_exists(field, field_id, fields))
+    {
+      // TODO: This could be a warning if the types match
+      error()
+        << "Field definition for " << field_id
+        << " already loaded from overlay class" << eom;
+      continue;
+    }
     debug() << "Adding symbol:  field '" << field.name << "'" << eom;
     convert(*class_symbol, field);
+    POSTCONDITION(check_field_exists(field, field_id, fields));
   }
 
-  // now do methods
-  for(const auto &method : c.methods)
+  // Now do methods
+  std::set<irep_idt> overlay_methods;
+  for(auto overlay_class : overlay_classes)
+  {
+    for(const methodt &method : overlay_class.get().methods)
+    {
+      const irep_idt method_identifier =
+        qualified_classname + "." + id2string(method.name)
+          + ":" + method.descriptor;
+      if(method_bytecode.contains_method(method_identifier))
+      {
+        // This method has already been discovered while parsing an overlay
```

Comment:
Sorry I realise you've tried to explain this more thoroughly, but I still don't get why `overlay_methods.count(method_identifier) == 0` is equivalent to the overlay method existing later in the classpath? Surely the first method on the first overlay class we see, overlay_methods will be empty and therefore this will be true. Perhaps I don't know what `method_bytecode` is?

Reply:
Added more comment. If it's still not clear then come and ask me.